docs/man/man3/Bms.3
.TH "Bms" 3 "Sun Jan 15 2017" "Version 0.0" "RaceUp_ino_core" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Bms \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Bms\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBBms\fP ()"
.br
.ti -1c
.RI "template<class T > void \fBsendSegmentStatus\fP (String statusType, byte cell, byte segment, T value) const "
.br
.ti -1c
.RI "template<class T > void \fBsendUpdate\fP (byte typeValue, T value) const "
.br
.ti -1c
.RI "template<class T > void \fBsendStatus\fP (T value) const "
.br
.ti -1c
.RI "void \fBsendSegmentVoltage\fP (byte cell, byte segment, double value) const "
.br
.ti -1c
.RI "void \fBsendPackVoltageOfBmsDevice\fP (byte n) const "
.br
.ti -1c
.RI "void \fBsendSegmentTemperature\fP (byte cell, byte segment, double value) const "
.br
.ti -1c
.RI "void \fBsendPackTemperatureOfBmsDevice\fP (byte n) const "
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static String \fBgetSerialUpdate\fP ()"
.br
.ti -1c
.RI "static String \fBcommandToString\fP (int command)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. \fBBms\fP device on entire battery pack 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Bms::Bms ()"
New bms 
.PP
\fBReturns:\fP
.RS 4
new bms 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "String Bms::commandToString (int command)\fC [static]\fP"
Translate enum command to a string 
.PP
\fBParameters:\fP
.RS 4
\fIcommand\fP command number (as from the previously defined enum) 
.RE
.PP
\fBReturns:\fP
.RS 4
string representing name of command 
.RE
.PP

.SS "String Bms::getSerialUpdate ()\fC [static]\fP"
Reads serial data and checks if there is an update coming from the client 
.PP
\fBReturns:\fP
.RS 4
Raw String read from serial 
.RE
.PP

.SS "void Bms::sendPackTemperatureOfBmsDevice (byte n) const"
Send temperature data on the pack of bmsDevice device 
.PP
\fBParameters:\fP
.RS 4
\fIn\fP bmsDevice device 
.RE
.PP

.SS "void Bms::sendPackVoltageOfBmsDevice (byte n) const"
Send voltage data on the pack of bmsDevice device 
.PP
\fBParameters:\fP
.RS 4
\fIn\fP bmsDevice device 
.RE
.PP

.SS "template<class T > void Bms::sendSegmentStatus (String statusType, byte cell, byte segment, T value) const\fC [inline]\fP"
Send segment data 
.PP
\fBParameters:\fP
.RS 4
\fIstatusType\fP String about type of status to broadcast 
.br
\fIcell\fP number of cell 
.br
\fIsegment\fP number of segment 
.br
\fIvalue\fP value to send 
.RE
.PP

.SS "void Bms::sendSegmentTemperature (byte cell, byte segment, double value) const"
Send temperature data on the segment 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIcell\fP Cell number 
.br
\fIsegment\fP Segment number 
.br
\fIsegmentVoltage\fP Temperature of segment in cell number 
.br
\fIvalue\fP value to send 
.RE
.PP

.SS "void Bms::sendSegmentVoltage (byte cell, byte segment, double value) const"
Send segment voltage data 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIcell\fP Cell number 
.br
\fIsegment\fP Segment number 
.br
\fIsegmentVoltage\fP Voltage of segment in cell number 
.br
\fIvalue\fP value to send
.RE
.PP
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. 
.SS "template<class T > void Bms::sendStatus (T value) const\fC [inline]\fP"
Send status 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIvalue\fP Status to send 
.RE
.PP

.SS "template<class T > void Bms::sendUpdate (byte typeValue, T value) const\fC [inline]\fP"
Send data 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fItypeValue\fP Type of arg to send 
.br
\fIvalue\fP value to send 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for RaceUp_ino_core from the source code\&.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.2.1 - 2017-01-15
### Fixed
- non-OOP residue includes

## 0.2 - 2017-01-15
### Refactored
- OOP design implemented
### Removed
- useless classes like BmsDeviceReader, BmsDeviceWriter BmsDeviceUpdater

## 0.1.3 - 2017-01-11
### Added
- utils frame conversions

## 0.1.2 - 2017-01-11
### Added
- SdCardStream
### Tested
- compilation with Arduino UNO, DUE

## 0.1.1 - 2017-01-11
### Fixed
- ```#include``` bugs

## 0.1 - 2017-01-11
### Added
- Utils, BmsDevice, BmsDeviceReader, BmsDeviceUpdater, BmsDeviceWriter classes
- documentation
- data

## 0.0 - 2017-01-11
### Added
- project structure and repository


src/Bms/Bms.cpp
/**
 * Copyright 2016 - 2017 RaceUp Team Unipd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "Bms.h"

void Bms::sendSegmentVoltage(byte
cell,
byte segment,
double value
) const {
sendSegmentStatus(commandToString(kCellVoltage), cell, segment, value
);
}

void Bms::sendPackVoltageOfBmsDevice(byte
n) const {
byte segment = (n - 1) / 3;
for (
byte cell = 0;
cell < 18; cell++) {  // loop through cells monitored by bms device (6 of them)
sendSegmentVoltage(cell, segment, BmsDevice(n)
.
getCellVoltage(cell)
);  // send voltage
}
}

void Bms::sendSegmentTemperature(byte
cell,
byte segment,
double value
) const {
sendSegmentStatus(commandToString(kCellTemperature), cell, segment, value
);
}

void Bms::sendPackTemperatureOfBmsDevice(byte
n) const {
byte segment = (n - 1) / 3;
for (
byte cell = 0;
cell < 18; cell++) {  // loop through cells monitored by bms device (6 of them)
sendSegmentTemperature(cell, segment, BmsDevice(n)
.
getTemperature(TEMPERATURE1)
);
}
}

String Bms::getSerialUpdate() {
    return Serial.readString();
}

String Bms::commandToString(int command) {
    switch (command) {
        case 0:
            return (String) "Acknowledge";
        case 1:
            return (String) "Error";
        case 2:
            return (String) "Log";
        case 3:
            return (String) "Plot";
        case 4:
            return (String) "Alert";
        case 5:
            return (String) "Fault";
        case 6:
            return (String) "Status";
        case 7:
            return (String) "Voltage";
        case 8:
            return (String) "Temperature";
        case 9:
            return (String) "Sleep";
        default:
            return (String) "Unknown";
    }
}

Bms::Bms() {}
